ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::ffi::c_void;
+use std::iter::once;
+
+use deno_core::v8::fast_api;
+use dynasmrt::dynasm;
+use dynasmrt::DynasmApi;
+use dynasmrt::ExecutableBuffer;
+
+use crate::NativeType;
+use crate::Symbol;
+
+pub(crate) fn is_compatible(sym: &Symbol) -> bool {
+  cfg!(all(target_arch = "x86_64", target_family = "unix"))
+    && !sym.can_callback
+    && !sym.parameter_types.iter().any(|t| !is_fast_api_arg(*t))
+    && is_fast_api_rv(sym.result_type)
+}
+
+pub(crate) fn compile_trampoline(sym: &Symbol) -> Trampoline {
+  Trampoline::compile(sym)
+}
+
+pub(crate) fn make_template(sym: &Symbol, trampoline: &Trampoline) -> Template {
+  let args = once(fast_api::Type::V8Value)
+    .chain(sym.parameter_types.iter().map(|t| t.into()))
+    .collect::<Vec<_>>();
+
+  Template {
+    args: args.into_boxed_slice(),
+    ret: (&fast_api::Type::from(&sym.result_type)).into(),
+    symbol_ptr: trampoline.ptr(),
+  }
+}
+
+/// Trampoline for fast-call FFI functions
+///
+/// Removes first argument (Javascript object receiver) and shifts the rest of arguments to the left
+pub(crate) struct Trampoline(ExecutableBuffer);
+
+impl Trampoline {
+  fn ptr(&self) -> *const c_void {
+    &self.0[0] as *const u8 as *const c_void
+  }
+
+  fn compile(sym: &Symbol) -> Self {
+    #[derive(Clone, Copy)]
+    struct ArgCounter {
+      integer: i32,
+      sse: i32,
+    }
+    let mut counter = ArgCounter {
+      integer: -6,
+      sse: -8,
+    };
+    let args_to_shift: Vec<_> = sym
+      .parameter_types
+      .iter()
+      .filter_map(|p| match p {
+        NativeType::F32 | NativeType::F64 => {
+          counter.sse += 1;
+          if counter.sse > 0 && counter.integer >= 0 {
+            // floats are only shifted to accomodate integer shift in the stack
+            Some(counter)
+          } else {
+            None
+          }
+        }
+        _ => {
+          counter.integer += 1;
+          Some(counter)
+        }
+      })
+      .collect();
+
+    let mut ops = dynasmrt::x64::Assembler::new().unwrap();
+    if args_to_shift.is_empty() {
+      dynasm!(ops
+        ; .arch x64
+        ; xor rdi, rdi
+      );
+    } else {
+      for arg in args_to_shift {
+        match arg {
+          ArgCounter { integer: -5, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdi, rsi
```

Comment:
Is the plain `mov` is not enough for 8 and 16 bit integers? The incoming call will call the trampoline using 32 bit integers, but the actual C function should called with the expected 8 or 16 bit integers.

I would imagine that eg. `-4` in u8 would look different from `-4` in u32 due to the sign bit but based on the tests that does not seem to be the case?

Reply:
In two's complement representation of signed integers, `-4i32` is just a "sign-extended" `-4i8`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=69e79c84d944641ec063c6d179d0aa42

Actually,[ it is a de-facto standard of most compilers](https://stackoverflow.com/questions/36706721/is-a-sign-or-zero-extension-required-when-adding-a-32bit-offset-to-a-pointer-for/36760539#36760539) to [sign-extend (or zero-extend in the case of unsigned) the arguments to fill in the lower 32 bits of the registers](https://godbolt.org/z/daaWvP8sr). In a way, regardless of V8, the functions are always getting an i32 anyway 😆

Reply:
Ah, I was not aware that the sign bit was always extended like that, I thought that sign extending was only about extending the sign bit up to the 32nd bit.

Still, this doesn't mean that it's okay to use a 32 bit integer as an 8 or 16 bit integer. As you said: Sign / zero extending is usually expected and even when it isn't, it doesn't hurt but we're not doing that here. So if a user calls a C function that expects a `u8` using eg. the integer value `3500` then this code will not do one blessed thing to that u32 value, ie. it will not zero-out the bits from 9 to 32, see: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=69e79c84d944641ec063c6d179d0aa42

Since we're our own compiler here, we must make sure to zero-out / sign-extend the top bits so that the de-facto standard is honored.

Reply:
Also, the Godbolt example you linked clearly does not use plain `mov` for the i64 to i16 call case but instead uses `movsx` to sign-extend so as to truncate the winder integer into a shorter one.

As another example, try this: https://godbolt.org/z/1d67vTbrW
Here "foo" and "bar" are like the trampolines that you're currently creating while the "foo_u8" / "bar_i16" are the actual foreign library functions. The trampolines call into a function that takes a shorter integer than their argument is, and thus they use `movzx` and `movsx` respectively to fix the mismatch.

Also note that the return value is not just plainly returned as is: There's a zero or sign extension done because again, the return value is smaller than what the trampoline should return and unlike the de-facto standard for sign extending arguments, there is no such standard on return values. A `u8` return value may well contain garbage bits in bits 9 to 32 and it is up to the receiver of that value to handle that if they want to it as a larger width integer.

Reply:
> Since we're our own compiler here, we must make sure to zero-out / sign-extend the top bits so that the de-facto standard is honored.

> A u8 return value may well contain garbage bits in bits 9 to 32 and it is up to the receiver of that value to handle that if they want to it as a larger width integer

Yep, you are right. 

Pending:
1. sign/zero extend 8 and 16 bit integer arguments up to 32 bits
2. sign/zero extend 8 and 16 bit integer return values up to the size of the trampoline return value (32 bits)

Reply:
Correction implemented. Added a couple of tests that corroborated the bug. Naturally, the implementation has complicated substantially, but overall I don't think it's too worrisome (yet).